Xplat/src/main/java/vazkii/botania/client/integration/shared/OrechidUIHelper.java

```diff
@@ -0,0 +1,171 @@
+package vazkii.botania.client.integration.shared;
+
+import it.unimi.dsi.fastutil.ints.IntIntImmutablePair;
+import it.unimi.dsi.fastutil.ints.IntIntPair;
+
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.Minecraft;
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.chat.Component;
+import net.minecraft.util.Mth;
+import net.minecraft.world.item.crafting.RecipeType;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import vazkii.botania.api.recipe.OrechidRecipe;
+import vazkii.botania.common.handler.OrechidManager;
+
+import java.text.NumberFormat;
+import java.util.stream.Stream;
+
+/**
+ * Shared helper methods for the various recipe display mod plugins.
+ */
+public class OrechidUIHelper {
+	/**
+	 * How far off from the actual chance an approximated outputs/inputs ratio calculated by
+	 * {@link #getRatioForChance(double)} should be at most.
+	 */
+	private static final float MAX_ACCEPTABLE_RATIO_ERROR = 0.05f;
+
+	/**
+	 * How many input blocks an approximated outputs/inputs ratio calculated by
+	 * {@link #getRatioForChance(double)} should have at most if the
+	 * number of outputs in the ratio is greater than 1.
+	 */
+	private static final int MAX_NUM_INPUTS_FOR_RATIO = 16;
+
+	@NotNull
+	public static Component getPercentageComponent(double chance) {
+		final String chanceText = LocaleHelper.formatAsPercentage(chance, 1);
+		return Component.literal(chanceText);
+	}
+
+	@NotNull
+	public static Component getRatioTooltipComponent(@NotNull IntIntPair ratio) {
+		final NumberFormat formatter = LocaleHelper.getIntegerFormat();
+		return Component.translatable("botaniamisc.conversionRatio",
+				formatter.format(ratio.secondInt()),
+				formatter.format(ratio.firstInt()));
+	}
+
+	@NotNull
+	public static Component getBiomeChanceTooltipComponent(double chance, @NotNull String biomeTranslationKey) {
+		final var ratio = getRatioForChance(chance);
+		return Component.translatable("botaniamisc.conversionChanceBiome",
+				getPercentageComponent(chance),
+				Component.translatable(biomeTranslationKey).withStyle(ChatFormatting.ITALIC),
+				getRatioTooltipComponent(ratio)
+		).withStyle(ChatFormatting.GRAY);
+	}
+
+	@Nullable
+	public static Stream<Component> getBiomeChanceAndRatioTooltipComponents(double chance, OrechidRecipe recipe) {
+		final var biomeTranslationKey = OrechidUIHelper.getPlayerBiomeTranslationKey();
+		final var player = Minecraft.getInstance().player;
+		if (biomeTranslationKey == null || player == null) {
+			return null;
```

Comment:
I would just return Stream.empty() from this method to make the result NotNull and avoid the long ternaries at the callsites (you could just blindly concat the two streams)